Remove banned functions

janvorli opened the pull request:
This change removes _snwprintf, _snprintf and _vsnwprintf usage from CoreCLR and
their implementations from PAL.
PAL exposes their secure variants instead and CoreCLR now uses those instead.
I have also removed the StringCchPrintfA/W, StringCchVPrintfA/W, StringCbVPrintfA/W,
StringCbPrintfA/W, StringCbPrintfExA/W, StringCchVPrintfExA/W, StringCbVPrintfExA/W
and StringCchPrintfExA/W replaced their usage by the secure variants of the sprintf
functions, since they were used at only few places and implementing all of the variants
using the secure sprintf variants would be a hassle.
I also needed to fix a missing support for size modifiers for %p formatting character
and for wide characters / strings in the secure sprintf functions that was revealed
by the PAL tests.
I have also removed a bunch of PAL tests that were using %n formatting character which
was not implemented since it is considered unsafe and translated PAL tests that were using
the removed functions to use the safe variants of those.

janvorli commented:
@jkotas can you take a look please? With this change, only the memcpy and strlen function from the list of banned functions are still used by coreclr.
Github seems to have hard time diffing the strsafe.h for some strange reason, I've only removed the printing functions from it. So when the diff starts looking weird, you can ignore it till the end of the file.

jkotas commented on src/pal/src/misc/perftrace.cpp:
```diff
@@ -33,7 +33,7 @@ Module Name:
 #include <errno.h>
 
-#define snprintf _snprintf
+#define snprintf _snprintf_s
```
>snprintf [](start = 8, length = 8)

Is there going to be compile error if PAL_PERF is ever enabled? I guess it does not matter since you are planning to delete this file?

janvorli replied:
@jkotas thanks for spotting this. Yes, I was going to change it and then decided I will remove this file later since it is of no future use for CoreCLR, but forgotten to revert the change here.

jkotas commented:
`src/vm/compactlayoutwriter.cpp` with no delta is part of the commit for some reason.

jkotas commented:
@rahku @ramarag Could you please review this as well?

rahku commented:
I didn't go through all of it. But the ones I went through looked good to me.

janvorli commented:
@jkotas I've fixed the OSX build issue, removed the empty src/vm/compactlayoutwriter.cpp that got accidentally in and also removed the change in the src/pal/src/misc/perftrace.cpp.
